File: find_first_duplicate.js
function findFirstDuplicate(arr) {
  // Implementation of finding the first duplicate
}

console.log("Expecting: 5");
test('can handle an array containing no duplicates', () => {
  expect(findFirstDuplicate([1, 2, 3, 4, 5])).toBe(-1);
});

console.log("Expecting: 4");
console.log("=>", findFirstDuplicate([2, 1, 4, 4, 2]));

console.log("");

console.log("Expecting: 4");
console.log("=>", findFirstDuplicate([1, 2, 3, 4]));



//Pseudocode: Create a functon named 'findFirstDuplicate' that takes an 'arr' as input.
//Create an empty object named 'seen'.
//iterate through each element 'num' in the array 'arr'.
//Check if 'num' is  already present in the  'seen' object.
//if  it is , return 'num' as a  key in the sen 'seen' object with with a value of 'true'.
//if no duplicate is  found, return --1.



// Please add your pseudocode to this file
// And a written explanation of your solution


//Explanation:
//The 'findFirstDuplicate' function checks for the first duplicate element in an array it  uses an object  
//called 'seen' to keep track of the elements that have  been so far . By iterating  th
//amd the function returns that element. If no duplicate is found, the function returns -1.
///

// Pseudocode
//Rewrite the promblem in your own words
//Validate that you understand the  problem
//Write your own tests
//Pseudocode: functionFirstDuplicate locates the first duplicate (arr) when the duplicates are not next towards one another', () => {
//Pseudocode: functionFirstDuiplicate  locates the Second duplicate ([1, 2, 3, 4, 5])).toBe( -1 )
//Code!

File: find_first_duplicate.py
def find_first_duplicate(arr):
    unique_nums = set()

    for num in arr:
        if num in unique_nums:
            return num
        unique_nums.add(num)

    return -1

print("Expecting: -2")
print(find_first_duplicate([0, 1, 2, 2, 4]))

print("Expecting: 4")
print(find_first_duplicate([1, 2, 4, 4, 5, 10]))

print("Expecting: -1")
print(find_first_duplicate([2]))

print("Expecting: -1")
print(find_first_duplicate([4]))

print("Expecting: 22")
print(find_first_duplicate([10, 33, 22, 70, 100]))


find_first_duplicate function:

function find_first_duplicate(arr):
    unique_nums = set()  #create an empty set to store unique numbers

    for num in arr:  # iterate through each number in the array
        if num in unique_nums:  # check if the number is already in the set
            return num  # if it is, return the number as the first duplicate
        unique_nums.add(num)  # if not, add the number to the set

    return -1  # if no duplicate is found, return -1